src/piton/featurizers/__init__.py

```diff
@@ -0,0 +1,115 @@
+        self.current_count: int = current_count
+        self.current_mean: float = current_mean
+        if current_count == 0 and current_variance == 0:
+            self.current_M2 = 0.0
+        elif current_count > 0:
+            self.current_M2 = current_variance * (current_count - 1)
+        else:
+            raise ValueError(
+                "Cannot specify `current_variance` with a value > 0"
+                "without specifying `current_count` with a value > 0."
+            )
+
+    def add(self, newValue: float) -> None:
+        """
+        Add an observation to the calculation using Welford's online algorithm.
+
+        Taken from: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
+        """
+        self.current_count += 1
+        delta: float = newValue - self.current_mean
+        self.current_mean += delta / self.current_count
+        delta2: float = newValue - self.current_mean
+        self.current_M2 += delta * delta2
+
+    def mean(self) -> float:
+        """
+        Return the current mean.
+        """
+        return self.current_mean
+
+    def variance(self) -> float:
+        """
+        Return the current sample variance.
+        """
+        if self.current_count < 2:
+            raise ValueError(
+                f"Cannot compute variance with only {self.current_count} observations."
+            )
+
+        return self.current_M2 / (self.current_count - 1)
+
+    def standard_deviation(self) -> float:
+        """
+        Return the current standard devation.
+        """
+        return math.sqrt(self.variance())
+
+    @classmethod
+    def merge_pair(
+        cls, stats1: OnlineStatistics, stats2: OnlineStatistics
+    ) -> OnlineStatistics:
+        """
+        Merge two sets of online statistics using Chan's parallel algorithm.
+
+        Taken from: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Parallel_algorithm
+        """
+        if stats1.current_count == 0:
+            return stats2
+        elif stats2.current_count == 0:
+            return stats1
+
+        count: int = stats1.current_count + stats2.current_count
+        delta: float = stats2.current_mean - stats1.current_mean
+        mean: float = stats1.current_mean + delta * stats2.current_count / count
+        M2 = (
+            stats1.current_M2
+            + stats2.current_M2
+            + delta**2 * stats1.current_count * stats2.current_count / count
+        )
+        return OnlineStatistics(count, mean, M2 / (count - 1))
+
+    @classmethod
+    def merge(cls, stats_list: List[OnlineStatistics]) -> OnlineStatistics:
+        """
+        Merge a list of online statistics.
+        """
+        if len(stats_list) == 0:
+            raise ValueError("Cannot merge an empty list of statistics.")
+        unmerged_stats = copy.deepcopy(stats_list)
+        stats1: OnlineStatistics = unmerged_stats.pop(0)
+        while len(unmerged_stats) > 0:
```

Comment:
You are supposed to run the merging algorithm in a tree like structure to get it to perform optimally.

Given the array [a, b, c, d]

ab = merge(a, b)
cd = merge(c, d)

result = merge(ab, cd)

Should be a pretty simple change to the code and will give a lot better performance.

Also, we really should have unit tests for this class.